operator/watchers/cilium_node_gc.go

```diff
@@ -0,0 +1,133 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// ciliumNodeGCCandidate keeps track of cilium nodes, which are candidate for GC.
+// Underlying there is a map with node name as key, and last marked timestamp as value.
+type ciliumNodeGCCandidate struct {
+	lock          lock.RWMutex
+	nodesToRemove map[string]time.Time
+}
+
+func newCiliumNodeGCCandidate() *ciliumNodeGCCandidate {
+	return &ciliumNodeGCCandidate{
+		nodesToRemove: map[string]time.Time{},
+	}
+}
+
+func (c *ciliumNodeGCCandidate) Get(nodeName string) (time.Time, bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	val, exists := c.nodesToRemove[nodeName]
+	return val, exists
+}
+
+func (c *ciliumNodeGCCandidate) Add(nodeName string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.nodesToRemove[nodeName] = time.Now()
+}
+
+func (c *ciliumNodeGCCandidate) Delete(nodeName string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	delete(c.nodesToRemove, nodeName)
+}
+
+// RunCiliumNodeGC performs garbage collector for cilium node resource
+func RunCiliumNodeGC(ctx context.Context, ciliumNodeStore cache.Store, interval time.Duration, stopCh <-chan struct{}) {
+	nodesInit(k8s.WatcherClient(), stopCh)
+
+	// wait for k8s nodes synced is done
+	select {
+	case <-slimNodeStoreSynced:
+	case <-ctx.Done():
+		return
+	}
+
+	log.Info("Starting to garbage collect stale CiliumNode custom resources")
+
+	candidateStore := newCiliumNodeGCCandidate()
+	// create the controller to perform mark and sweep operation for cilium nodes
+	ctrlMgr.UpdateController("cilium-node-gc",
+		controller.ControllerParams{
+			Context: ctx,
+			DoFunc: func(ctx context.Context) error {
+				return performCiliumNodeGC(ctx, k8s.CiliumClient().CiliumV2().CiliumNodes(), ciliumNodeStore,
+					nodeGetter{}, interval, candidateStore)
+			},
+			RunInterval: interval,
+		},
+	)
+}
+
+func performCiliumNodeGC(ctx context.Context, client ciliumv2.CiliumNodeInterface, ciliumNodeStore cache.Store,
+	nodeGetter slimNodeGetter, interval time.Duration, candidateStore *ciliumNodeGCCandidate) error {
+	for _, nodeName := range ciliumNodeStore.ListKeys() {
+		scopedLog := log.WithField(logfields.NodeName, nodeName)
+		_, err := nodeGetter.GetK8sSlimNode(nodeName)
+		if err == nil {
+			scopedLog.Debugf("CiliumNode is valid, no gargage collection required")
+			continue
+		}
+
+		if !k8serrors.IsNotFound(err) {
+			scopedLog.WithError(err).Error("Unable to fetch k8s node from store")
+			return err
+		}
+
+		obj, _, err := ciliumNodeStore.GetByKey(nodeName)
+		if err != nil {
+			scopedLog.WithError(err).Error("Unable to fetch cilium node from store")
```

Comment:
Nit: there's a mix of usages between "ciilum node" and "CiliumNode". I suggest we stick with the latter everywhere.